bridge-worker/btc_withdrawal.go

```diff
@@ -0,0 +1,632 @@
+				// We need to start over the event processing as the data
+				// needed for `withdrawBTC` might have changed in the meantime.
+				for i := 0; i < 5; i++ {
+					withdrawalProcessLogger := withdrawalLogger.With("iteration", i)
+
+					if i > 0 {
+						// use a backoff for subsequent iterations as they are
+						// most likely retries
+						select {
+						case <-time.After(withdrawalProcessBackoff):
+						case <-ctx.Done():
+							withdrawalProcessLogger.Warn(
+								"stopping withdrawal submission process backoff " +
+									"wait due to context cancellation",
+							)
+							return
+						}
+					}
+
+					// check if still pending
+					ok, err := bw.isPendingBTCWithdrawal(event)
+					if err != nil {
+						withdrawalProcessLogger.Error(
+							"failed to check if pending; retrying",
+							"error", err,
+						)
+						continue
+					}
+					if !ok {
+						withdrawalProcessLogger.Info(
+							"withdrawal no longer pending; skipping",
+						)
+						withdrawingSuccessful = true
+						break
+					}
+
+					withdrawalProcessLogger.Info(
+						"submitting Bitcoin withdrawal transaction",
+					)
+
+					tx, err := bw.mezoBridgeContract.WithdrawBTC(
+						entry,
+						walletPublicKeyHash,
+						mainUTXO,
+					)
+					if err != nil {
+						withdrawalProcessLogger.Error(
+							"withdrawal transaction submission failed; retrying",
+							"error", err,
+						)
+						continue
+					}
+
+					// schedule finality check
+					bw.queueWithdrawalFinalityCheck(event)
+
+					withdrawingSuccessful = true
+
+					withdrawalProcessLogger.Info(
+						"Bitcoin withdrawal submitted",
+						"tx_hash", tx.Hash().Hex(),
+					)
+					break
+				}
+
+				if !withdrawingSuccessful {
+					withdrawalLogger.Error(
+						"all withdrawal attempts failed; re-queuing",
+					)
+					bw.enqueueBtcWithdrawal(event)
+				}
+			}
+		case <-ctx.Done():
+			bw.logger.Warn(
+				"stopping Bitcoin withdrawal queue loop due to context " +
+					"cancellation",
+			)
+			return
+		}
+	}
+}
+
+// queueWithdrawalFinalityCheck constructs a Bitcoin withdrawal finality check
+// and puts it into a queue.
+func (bw *BridgeWorker) queueWithdrawalFinalityCheck(
+	event *portal.MezoBridgeAssetsUnlockConfirmed,
+) {
+	bw.withdrawalFinalityChecksMutex.Lock()
+	defer bw.withdrawalFinalityChecksMutex.Unlock()
+
+	check := &withdrawalFinalityCheck{
+		event:             event,
+		scheduledAtHeight: nil,
+	}
+
+	// This should never happen. Check just in case.
+	if _, ok := bw.withdrawalFinalityChecks[check.key()]; ok {
+		return
+	}
+
+	bw.withdrawalFinalityChecks[check.key()] = check
+
+	bw.logger.Info(
+		"queued Bitcoin withdrawal finality check",
+		"unlock_sequence", event.UnlockSequenceNumber.String(),
+	)
+}
+
+// processWithdrawalFinalityChecks selects Bitcoin withdrawal finality checks
+// that have reached their scheduled height and executes them.
+func (bw *BridgeWorker) processWithdrawalFinalityChecks(ctx context.Context) {
```

Comment:
Non-blocking: Let's improve nomenclature and prefix "withdrawal" with "BTC" everywhere, in the names, and in the log messages. We need to be crystal clear we are dealing with BTC withdrawals here.